CPU: fix branch page crossing + offset calulation

Holladiewal opened the pull request:
Page crossing and offset during a branch are computed from the next instruction onward, because the 6502 advances the PC during fetching of the opcode and operand

This means, that a branch at CFFE to D0xx doesnt count as crossing a page boundary. Advancement of the PC also causes the base address of the branch to shift, as the offset is not applied to where the branch opcode is, but where the PC is after reading the whole branch instruction.

I also changed how the PC is set to avoid casting the PC to i16 and back to u16, just to get rid of a potential headache with the high bit of the PC being set and potentially causing the i16 representation to be negative (depending on how jakt/c++ do the cast in the background). 

~~Since I didnt have jakt around when writing these changes, I'll mark the PR as a draft for now.~~
Real PR now since I had a chance to pass it through jakt and fix all the compiler errors.

Holladiewal committed:
CPU: fix branch page crossing + offset calulation

Page crossing and offset during a branch are computed from the next instruction onward, because the 6502 advances the PC during fetching of the opcode and operand

This means, that a branch at CFFE to D0xx doesnt count as crossing a page boundary. Advancement of the PC also causes the base address of the branch to shift, as the offset is not applied to where the branch opcode is, but where the PC is after reading the whole branch instruction.

jntrnr commented:
lgtm

sophiajt commented on src/cpu.jakt:
```diff
@@ -1419,13 +1419,29 @@ class CPU {
+        .pc += 2
+        
+        let offset: u8 = 0
+        let negative = false
```
does this ever change? I see it checked here but then doesn't seem to get updated

Holladiewal replied:
You are totally right, I forgot to update it whenever the offset was negative.
Should all be fixed now.

Holladiewal committed:
CPU: Fix not setting the negative flag

We introduced the negative variable for the offset calculation during branching, but never updated it.
This is not the negative register!

Holladiewal committed:
CPU: rename negative variable in branch()

Since this is not the actual negative flag but just a helper for us, lets rename it to lessen confusion between `negative` and `.negative`

Holladiewal committed:
CPU: Fix compiler errors

Mutable variables should be mutable, and the `-1` constant also
needs to be an i8.

jntrnr commented:
thanks!
